### ISSUE TITLE
Validation Splits Not Available

### ISSUE THREAD

**Issue by Zhou-Hangyu** [tanganke/fusion_bench#12]
Currently seems that your hf datasets don't have validation set and the codebase also does not support evaluate on val sets. But this is used in grid search for the optimal coefficients in task arithmetic. Such feature will be nice to have.

**Comment by tanganke** [comment]
I have added a [helper function](https://github.com/tanganke/fusion_bench/blob/da0eca115a8728aaf3307f2d3faf7dfc06fb2a32/fusion_bench/utils/data.py#L56C1-L94C66) to split the training dataset into training and validation sets. 

```python
def train_validation_split(
    dataset: Dataset,
    validation_fraction: float = 0.1,
    random_seed: Optional[int] = None,
    return_split: Literal["all", "train", "val"] = "both",
):
    """
    Split a dataset into a training and validation set.

    Args:
        dataset (Dataset): The dataset to split.
        validation_fraction (float): The fraction of the dataset to use for validation.
        random_seed (Optional[int]): The random seed to use for reproducibility.
        return_split (Literal["all", "train", "val"]): The split to return.

    Returns:
        Tuple[Dataset, Dataset]: The training and validation datasets.
    """
    num_samples = len(dataset)
    assert 0 < validation_fraction < 1, "Validation fraction must be between 0 and 1"
    generator = (
        torch.Generator().manual_seed(random_seed) if random_seed is not None else None
    )

    num_validation_samples = int(num_samples * validation_fraction)
    num_training_samples = num_samples - num_validation_samples
    training_dataset, validation_dataset = torch.utils.data.random_split(
        dataset, [num_training_samples, num_validation_samples], generator=generator
    )

    # return the split as requested
    if return_split == "all":
        return training_dataset, validation_dataset
    elif return_split == "train":
        return training_dataset
    elif return_split == "val":
        return validation_dataset
    else:
        raise ValueError(f"Invalid return_split: {return_split}")
```

A validation set can be specified in a config file. For example, for [SUN397](https://github.com/tanganke/fusion_bench/blob/da0eca115a8728aaf3307f2d3faf7dfc06fb2a32/config/dataset/image_classification/val/sun397.yaml#L1-L9):

```yaml
sun397:
  _target_: fusion_bench.utils.data.train_validation_split
  dataset:
    _target_: datasets.load_dataset
    path: tanganke/sun397
    split: train
  validation_fraction: 0.1
  random_seed: 0
  return_split: val
```

Use the validation splits to instantiate the task pool, the [task pool configuration](https://github.com/tanganke/fusion_bench/blob/da0eca115a8728aaf3307f2d3faf7dfc06fb2a32/config/taskpool/CLIPVisionModelTaskPool/clip-vit-classification_TA8_val.yaml#L1-L12) would be:

```yaml
defaults:
  - CLIPVisionModelTaskPool@: _template
  # use validation set as test set
  - /dataset/image_classification/val@test_datasets:
      - sun397
      - stanford-cars
      - resisc45
      - eurosat
      - svhn
      - gtsrb
      - mnist
      - dtd
```

Then a simple solution is to use the validation set as the test set and iterate over the scaling factor in a bash script. Here is an example:

```bash
for scaling_factor in $(seq 0.1 0.1 1.0); do
    echo "scaling_factor: ${scaling_factor}"
    fusion_bench \
        method=task_arithmetic \
        method.scaling_factor=${scaling_factor} \
        modelpool=CLIPVisionModelPool/clip-vit-base-patch32_TA8_model_only \
        taskpool=CLIPVisionModelTaskPool/clip-vit-classification_TA8_val \
        report_save_path=outputs/clip-vit-base-patch32/task_arithmetic/${scaling_factor}.json
done
```

**Comment by Zhou-Hangyu** [comment]
Thanks for the changes!
